images/controller/internal/controllers/rv_status_config_shared_secret/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	if maxFailedRVR != nil {
+		logFields = append(logFields,
+			"maxFailedIndex", maxFailedIndex,
+			"maxFailedRVR", maxFailedRVR.Name,
+			"maxFailedRVRNode", maxFailedRVR.Spec.NodeName,
+			"maxFailedAlgorithm", algorithms[maxFailedIndex],
+		)
+	} else {
+		logFields = append(logFields, "maxFailedIndex", maxFailedIndex)
+	}
+
+	if len(failedNodeNames) > 0 {
+		logFields = append(logFields, "failedNodes", failedNodeNames)
+	}
+
+	return logFields
+}
+
+// reconcileHandleUnsupportedAlgorithm checks RVRs for UnsupportedAlgorithm errors and switches to next algorithm
+func (r *Reconciler) reconcileHandleUnsupportedAlgorithm(
+	ctx context.Context,
+	rv *v1alpha3.ReplicatedVolume,
+	log logr.Logger,
+) (reconcile.Result, error) {
+	// Get all RVRs for this RV
+	rvrList := &v1alpha3.ReplicatedVolumeReplicaList{}
+	if err := r.cl.List(ctx, rvrList); err != nil {
+		log.Error(err, "Listing ReplicatedVolumeReplicas")
+		return reconcile.Result{}, err
+	}
+
+	// Collect all RVRs with errors
+	var rvrsWithErrors []*v1alpha3.ReplicatedVolumeReplica
+	var failedNodeNames []string
+	for _, rvr := range rvrList.Items {
+		if rvr.Spec.ReplicatedVolumeName != rv.Name {
+			continue
+		}
+		if HasUnsupportedAlgorithmError(&rvr) {
+			failedNodeNames = append(failedNodeNames, rvr.Spec.NodeName)
+			rvrsWithErrors = append(rvrsWithErrors, &rvr)
+		}
+	}
+
+	// If no errors found, nothing to do
+	if len(failedNodeNames) == 0 {
+		return reconcile.Result{}, nil
+	}
+
+	algorithms := v1alpha3.SharedSecretAlgorithms()
+
+	// Find maximum index among all failed algorithms and RVR with max algorithm
+	maxFailedIndex := -1
+	var maxFailedRVR *v1alpha3.ReplicatedVolumeReplica
+	var rvrsWithoutAlg []string
+	for _, rvr := range rvrsWithErrors {
+		// Access UnsupportedAlg directly, checking for nil
+		var unsupportedAlg string
+		if rvr.Status != nil && rvr.Status.DRBD != nil && rvr.Status.DRBD.Errors != nil &&
+			rvr.Status.DRBD.Errors.SharedSecretAlgSelectionError != nil {
+			unsupportedAlg = rvr.Status.DRBD.Errors.SharedSecretAlgSelectionError.UnsupportedAlg
+		}
+
+		if unsupportedAlg == "" {
+			rvrsWithoutAlg = append(rvrsWithoutAlg, rvr.Name)
+			continue
+		}
+
+		index := slices.Index(algorithms, unsupportedAlg)
+		if index != -1 && index > maxFailedIndex {
+			maxFailedIndex = index
+			maxFailedRVR = rvr
+		}
+	}
+
+	// If no valid algorithms found in errors (all empty), we cannot determine which algorithm is unsupported
+	// Log this issue and do nothing - we should not switch algorithm without knowing which one failed
+	if maxFailedIndex == -1 {
+		logFields := []interface{}{"rv", rv.Name, "failedNodes", failedNodeNames}
+		if len(rvrsWithoutAlg) > 0 {
+			logFields = append(logFields, "rvrsWithoutAlg", rvrsWithoutAlg)
+		}
+		log.V(1).Info("UnsupportedAlg is empty for all failed RVRs, cannot determine which algorithm to switch", logFields...)
+		return reconcile.Result{}, nil // Do nothing - we don't know which algorithm is unsupported
+	}
```

Comment:
Он сделан под такой паттерн:

```suggestion
	if maxFailedIndex == -1 {
		log := log.WithValues("rv", rv.Name, "failedNodes", failedNodeNames)
		if len(rvrsWithoutAlg) > 0 {
			log = log.WithValues("rvrsWithoutAlg", rvrsWithoutAlg)
		}
		log.V(1).Info("UnsupportedAlg is empty for all failed RVRs, cannot determine which algorithm to switch")
		return reconcile.Result{}, nil // Do nothing - we don't know which algorithm is unsupported
	}
```

```go
log:=log.WithValues() // сделать логгер с полями в скоупе - log из скоупа выше не меняется
log=log.WithValues() // привязать поля в логер parent scope
```